src/deepdisc/data_format/file_io.py

```diff
@@ -1,5 +1,192 @@
+
+        # Store the result in a class property for future use.
+        self.filedict = filenames_dict
+
+        return self
+
+    def generate_dataset_dict(self, func=None, filedict=None, filters=True, **kwargs):
+        """Generates a list of dictionaries using a user-defined annotation
+        generator function on each image file/mask. The format is determined
+        by the user defined function
+
+        Parameters
+        ----------
+        func: function
+            A user-defined function that operates on a set of images and a mask
+            file to generate a dictionary of annotations. The DataLoader
+            expects this function to take in kwargs as follows
+            (image_files, mask_file, **kwargs), where image files is a list of
+            paths to image filenames (each image corresponds to one band) and
+            mask_file points to a single mask filename.
+        filedict: dict
+            A dictionary with image and mask filepaths defined, generated by
+            `DataLoader.generate_filedict`. If not specified, attempts to use
+            a filedict stored within the `DataLoader` instance.
+        filters: bool
+            Determines whether the list of filters is passed along to the
+            annotation function. If true is passed along as
+            (images, mask, index, filters, other kwargs).
+
+        Returns
+        -------
+        self : DataLoader
+            A DataLoader with a dataset dictionary generated. Access using
+            `DataLoader.get_dataset()`.
+        """
+
+        if func is None:
+            raise ValueError(
+                "No annotation function has been provided. Please supply a function that takes in arguments: (image_files, mask_file, index, and optionally filters)."
+            )
+        if filedict is None:
+            if self.filedict is None:
+                raise ValueError("No file dictionary has been provided.")
+            else:
+                filedict = self.filedict
+
+        # Group images by filter
+        img_files = np.transpose([filedict[filt]["img"] for filt in filedict["filters"]])
+
+        # Initialize data dictionary
+        dataset_dicts = []
+
+        # Use user-provided function to generate a dictionary record per image set
+        for images, mask, index in zip(img_files, filedict["mask"], filedict["index"]):
+            # pass along filter list if requested
+            if filters:
+                record = func(images, mask, index, filedict["filters"], **kwargs)
+            else:
+                record = func(images, mask, index, **kwargs)
+
+            # Add records to the data_dict
+            dataset_dicts.append(record)
+
+        self.dataset = dataset_dicts
+        return self
+
+    def load_coco_json_file(self, file):
+        """Open a JSON text file, and return encoded data as dictionary.
+
+        Assumes JSON data is in the COCO format.
```

Comment:
Note for future work - is there a reasonable set of validation steps that we should perform here to verify that the loaded data is indeed in COCO format? Perhaps detectron2 offers some validation function???

Reply:
I'm not seeing anything obvious in detectron2... We could write something explicit that checks the key list contains the keys needed by detectron2?

Reply:
I did a little looking around too, and didn't find any packages that were immediately helpful that would do this for us. I'll create an issue for this so that we don't forget about it. Unless you think it would be pretty quick to write a validator, I would advocate for not putting a bunch of time into this one right now.

Reply:
https://github.com/lincc-frameworks/deepdisc/issues/65

Reply:
I'm in favor of holding off on this for now as well